### PR TITLE
hugetlb: Add support of Set and GetStats function

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao <mashimiao.fnst@cn.fujitsu.com>

### CODE REVIEW

**Comment by crosbymichael** [comment]
LGTM

**Comment by crosbymichael** [comment]
Do you know of a better way to represent this data instead of something like `2M:2312323`?  Maybe a struct instead of parsing the string?

**Comment by Mashimiao** [comment]
change it to struct is OK. But why do you think `2M:2312323` is not good?

**Comment by Mashimiao** [comment]
ping @LK4D4

**Comment by crosbymichael** [comment]
@Mashimiao just having type safety instead of string parsing in the API

**Comment by Mashimiao** [comment]
@crosbymichael
modified and rebased

File: configs/hugetlb.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package configs
+
+type Hugetlb struct {
```
Comment (vmarmol) [inline configs/hugetlb.go]: We should stick this one in the cgroups file since all the cgroup stuff is there

File: configs/hugetlb.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+	// which type of hugepage to limit.
+	Pagesize string `json:"page_size"`
```
Comment (vmarmol) [inline configs/hugetlb.go]: Isn't there a limited set of these? If so, what about an enum?

File: cgroups/stats.go [hunk 3]
```diff
@@ -63,13 +63,24 @@ type BlkioStats struct {
...
+}
+
 type Stats struct {
-	CpuStats    CpuStats    `json:"cpu_stats,omitempty"`
-	MemoryStats MemoryStats `json:"memory_stats,omitempty"`
-	BlkioStats  BlkioStats  `json:"blkio_stats,omitempty"`
+	CpuStats     CpuStats                `json:"cpu_stats,omitempty"`
+	MemoryStats  MemoryStats             `json:"memory_stats,omitempty"`
+	BlkioStats   BlkioStats              `json:"blkio_stats,omitempty"`
+	HugetlbStats map[string]HugetlbStats `json:"hugetlb_stats,omitempty"`
```
Comment (vmarmol) [inline cgroups/stats.go]: Can we document what this is a map of please? I'm guessing hugepage size to stats about that size?

**Comment by vmarmol** [comment]
Thanks for the PR @Mashimiao!

File: cgroups/stats.go (hunk 3 above)
Comment (Mashimiao) [inline cgroups/stats.go]: Yes, you are right. I will document it.

File: configs/hugetlb.go (hunk 1 above)
Comment (Mashimiao) [inline configs/hugetlb.go]: I think the name `Hugetlb` makes you confused.
We already have HugetlbGroup in cgroup/fs/hugetlb.go
I will change the name of struct and file

File: configs/hugetlb.go (hunk 2 above)
Comment (Mashimiao) [inline configs/hugetlb.go]: No, I think it's better to dynamically get what hugepage the host support

**Comment by Mashimiao** [comment]
ping @vmarmol

File: configs/hugetlb.go (hunk 2 above)
Comment (vmarmol) [inline configs/hugetlb.go]: Hmmm this type of stuff makes it a very hard API to consume. How would I know what values are allowed? How to I know which are available?

I'm OK starting with this, but can we at least put example of valid values? :)

File: configs/hugetlb.go (hunk 1 above)
Comment (vmarmol) [inline configs/hugetlb.go]: We should still put it in the cgroup file, the divisions of files there are very broad: cgroup, network, namespaces. This one is too fine to be in its own file.

File: cgroups/fs/hugetlb.go [hunk 4]
```diff
@@ -25,5 +40,37 @@ func (s *HugetlbGroup) Remove(d *data) error {
 }
 
 func (s *HugetlbGroup) GetStats(path string, stats *cgroups.Stats) error {
+	pageSizes, err := getHugePageSize()
```
Comment (vmarmol) [inline cgroups/fs/hugetlb.go]: Let's only do the getHugePageSize() once in create and cache that.

Comment (Mashimiao) [inline cgroups/fs/hugetlb.go]: Fixed.

Comment (vmarmol) [inline cgroups/fs/hugetlb.go]: I mean to do this once in a `NewHugetlbGroup()` so we only need to do it once per run of libcontainer, not at every stats collection. Or do the hugepage sizes change frequently?

File: configs/hugetlb.go (hunk 1 above)
Comment (vmarmol) [inline configs/hugetlb.go]: ping on this

File: configs/hugetlb.go (hunk 2 above)
Comment (vmarmol) [inline configs/hugetlb.go]: ping on adding example.

File: cgroups/fs/hugetlb.go (hunk 4 above)
Comment (Mashimiao) [inline cgroups/fs/hugetlb.go]: @vmarmol  Any XXXGroup is created When subsystems is created. And in Manager.Apply() used first. Do you mean put it into Manager.Apply()?

Comment (vmarmol) [inline cgroups/fs/hugetlb.go]: We can either make a New() and use that where we instantiate today: https://github.com/docker/libcontainer/blob/master/cgroups/fs/apply_raw.go#L26

Or we can probably do an init() function and save it as a global

**Comment by Mashimiao** [comment]
@vmarmol  I have modified code. WDYT?

**Comment by LK4D4** [comment]
LGTM
ping @vmarmol

**Comment by vmarmol** [comment]
Sorry for the delay. LGTM

**Pull request merged**